### PR TITLE
Fix warnings from clippy

### PR DESCRIPTION
**Pull request by zjp-CN**
I found clippy complained a lot and thus fix them.

<details>                                                                                                                                                                                         <summary>cargo clippy --workspace --all-targets</summary>                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           

```rust
helix $ cargo clippy --workspace --all-targets
warning: items were found after the testing module
   --> helix-loader\src\lib.rs:213:1
    |
213 | / mod merge_toml_tests {
214 | |     use std::str;
215 | |
216 | |     use super::merge_toml_values;
...   |
300 | |     (current_dir, true)
301 | | }
    | |_^
    |
    = help: move the items to before the testing module was defined
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#items_after_test_module
    = note: `#[warn(clippy::items_after_test_module)]` on by default

    Checking helix-core v0.6.0 (E:\Rust\helix\helix-core)
warning: `helix-loader` (lib test) generated 1 warning
   Compiling helix-term v0.6.0 (E:\Rust\helix\helix-term)
warning: taking a reference on `std::path::MAIN_SEPARATOR` conversion to `String`
   --> helix-core\src\syntax.rs:191:45
    |
191 |                         suffix.replace('/', &std::path::MAIN_SEPARATOR.to_string()),
    |                                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `std::path::MAIN_SEPARATOR_STR`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_main_separator_str
    = note: `#[warn(clippy::manual_main_separator_str)]` on by default

warning: `helix-core` (lib) generated 1 warning (run `cargo clippy --fix --lib -p helix-core` to apply 1 suggestion)
    Checking helix-lsp v0.6.0 (E:\Rust\helix\helix-lsp)
    Checking helix-vcs v0.6.0 (E:\Rust\helix\helix-vcs)
    Checking helix-dap v0.6.0 (E:\Rust\helix\helix-dap)
warning: useless conversion to the same type: `std::str::MatchIndices<'_, char>`
   --> helix-core\src\surround.rs:398:48
    |
398 |           let selections: SmallVec<[Range; 1]> = spec
    |  ________________________________________________^
399 | |             .match_indices('^')
400 | |             .into_iter()
    | |________________________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
    = note: `#[warn(clippy::useless_conversion)]` on by default
help: consider removing `.into_iter()`
    |
398 ~         let selections: SmallVec<[Range; 1]> = spec
399 +             .match_indices('^')
    |

warning: useless conversion to the same type: `std::str::MatchIndices<'_, char>`
   --> helix-core\src\surround.rs:404:40
    |
404 |           let expectations: Vec<usize> = spec
    |  ________________________________________^
405 | |             .match_indices('_')
406 | |             .into_iter()
    | |________________________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
help: consider removing `.into_iter()`
    |
404 ~         let expectations: Vec<usize> = spec
405 +             .match_indices('_')
    |

warning: `helix-core` (lib test) generated 3 warnings (1 duplicate) (run `cargo clippy --fix --lib -p helix-core --tests` to apply 2 suggestions)
    Checking helix-view v0.6.0 (E:\Rust\helix\helix-view)
warning: `Box::new(_)` of default value
  --> helix-view\src\clipboard.rs:71:5
   |
71 |     Box::new(provider::WindowsProvider::default())
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `Box::<WindowsProvider>::default()`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#box_default
   = note: `#[warn(clippy::box_default)]` on by default

warning: use of `default` to create a unit struct
  --> helix-view\src\clipboard.rs:71:39
   |
71 |     Box::new(provider::WindowsProvider::default())
   |                                       ^^^^^^^^^^^ help: remove this call to `default`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#default_constructed_unit_structs
   = note: `#[warn(clippy::default_constructed_unit_structs)]` on by default

warning: `helix-view` (lib) generated 2 warnings (run `cargo clippy --fix --lib -p helix-view` to apply 2 suggestions)
    Checking helix-tui v0.6.0 (E:\Rust\helix\helix-tui)
warning: casting to the same type is unnecessary (`u16` -> `u16`)
   --> helix-tui\src\buffer.rs:445:39
    |
445 |         let mut index = self.index_of(max_offset as u16, y);
    |                                       ^^^^^^^^^^^^^^^^^ help: try: `max_offset`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_cast
    = note: `#[warn(clippy::unnecessary_cast)]` on by default

warning: `helix-tui` (lib) generated 1 warning (run `cargo clippy --fix --lib -p helix-tui` to apply 1 suggestion)
warning: `helix-tui` (lib test) generated 1 warning (1 duplicate)
error: this looks like you are trying to swap `l2` and `tree.focus`
   --> helix-view\src\tree.rs:731:9
    |
731 | /         let l2 = tree.focus;
732 | |
733 | |         // Tree in test
734 | |         // | L0  | L2 |    |
735 | |         // |    L1    | R0 |
736 | |         tree.focus = l2;
    | |_______________________^ help: try: `std::mem::swap(&mut l2, &mut tree.focus)`
    |
    = note: or maybe you should use `std::mem::replace`?
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#almost_swapped
    = note: `#[deny(clippy::almost_swapped)]` on by default

warning: `helix-view` (lib test) generated 2 warnings (2 duplicates)
error: could not compile `helix-view` (lib test) due to previous error; 2 warnings emitted
warning: build failed, waiting for other jobs to finish...
warning: casting to the same type is unnecessary (`usize` -> `usize`)
   --> helix-term\src\ui\editor.rs:106:28
    |
106 |             let dap_line = frame.line.saturating_sub(1) as usize;
    |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `frame.line.saturating_sub(1)`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_cast
    = note: `#[warn(clippy::unnecessary_cast)]` on by default

warning: `helix-term` (lib) generated 1 warning (run `cargo clippy --fix --lib -p helix-term` to apply 1 suggestion)
warning: `helix-term` (lib test) generated 1 warning (1 duplicate)
```

</details>

### CODE REVIEW

**Comment by zjp-CN** [comment]
[std::path::MAIN_SEPARATOR_STR](https://doc.rust-lang.org/stable/std/path/constant.MAIN_SEPARATOR_STR.html) isn't stable on 1.65 and stable from [1.68](https://releases.rs/docs/1.68.0/#stabilized-apis) 

I'll update the PR by using the project toolchain instead of my local nightly rustc.

File: helix-loader/src/lib.rs [hunk 1]
```diff
@@ -281,21 +299,3 @@ mod merge_toml_tests {
...
-pub fn find_workspace() -> (PathBuf, bool) {
-    let current_dir = std::env::current_dir().expect("unable to determine current directory");
-    for ancestor in current_dir.ancestors() {
-        if ancestor.join(".git").exists() || ancestor.join(".helix").exists() {
-            return (ancestor.to_owned(), false);
-        }
-    }
-
-    (current_dir, true)
-}
```
Comment (archseer) [inline helix-loader/src/lib.rs:301]: Why the change here?

Comment (zjp-CN) [reply helix-loader/src/lib.rs:301]: Clippy says the test items should avoid be declared after testing mod: https://rust-lang.github.io/rust-clippy/master/index.html#items_after_test_module

So I just move the testing mod down, but git diff shows the changes on `find_workspace` instead.

```rust
warning: items were found after the testing module
   --> helix-loader\src\lib.rs:213:1
    |
213 | / mod merge_toml_tests {
214 | |     use std::str;
215 | |
216 | |     use super::merge_toml_values;
...   |
300 | |     (current_dir, true)
301 | | }
    | |_^
    |
    = help: move the items to before the testing module was defined
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#items_after_test_module
    = note: `#[warn(clippy::items_after_test_module)]` on by default
```

Comment (zjp-CN) [reply helix-loader/src/lib.rs:301]: Oh, I think the reason why I meet clippy complaints is that I use a rustc/cargo in a higher verion.

`items_after_test_module` hasn't been in [clippy1.65](https://rust-lang.github.io/rust-clippy/rust-1.65.0/) yet.

So the Github action doesn't catch them for now.

If this PR is not appropriate for now, feel free to close.

Comment (archseer) [reply helix-loader/src/lib.rs:301]: Yeah all these are future lints that don't trigger in current version yet but it's good to fix them ahead of time 👍🏻
